File: configurator/src/shared/led-paint.ts
import { rampColor } from './color-ramp'
import { rangeFraction } from './telemetry-value'
import { colorsOf, newRuleState, type LayerColors, type LedRuleState } from './led-colors'
import {
  OFF,
  areaOf,
  deviceShape,
  isMatrix,
  lampsOf,
  matrixLamp,
  maskHolds,
  type LampArea,
  type MatrixShape
} from './led-render'
import { paintSprite, paintText, panelOf } from './led-matrix-paint'
import type { HardwareDeviceConfiguration, LedEffect, RgbColor } from './configuration-schema'

const DEFAULT_BRIGHTNESS = 128
const FULL_BRIGHTNESS = 255

export interface PaintInput {
  value: number
  elapsedMs: number
  gates: readonly boolean[]
  valueText?: readonly (string | undefined)[]
  watched?: readonly (number | undefined)[]
  values?: readonly (number | undefined)[]
  states?: readonly LedRuleState[]
}

function fillArea(
  frame: RgbColor[],
  device: HardwareDeviceConfiguration,
  area: LampArea,
  color: RgbColor
): void {
  const shape = deviceShape(device)
  for (let row = 0; row < area.height; ++row) {
    for (let column = 0; column < area.width; ++column) {
      if (!maskHolds(area, column, row)) continue
      const lamp = matrixLamp(shape, area.x + column, area.y + row)
      if (lamp >= 0 && lamp < frame.length) frame[lamp] = color
    }
  }
}

function rawValueOf(effect: LedEffect, sweep: number): number {
  const minimum = effect.minimum ?? 0
  const maximum = effect.maximum ?? 1
  return minimum + sweep * (maximum - minimum)
}

function valueOf(
  effect: LedEffect,
  live: number | undefined,
  sweep: number
): number | undefined {
  if ((effect.source?.binding ?? '') === '') return undefined
  return live ?? rawValueOf(effect, sweep)
}

function wholeText(value: number): string {
  return String(Math.trunc(value < 0 ? value - 0.5 : value + 0.5))
}

interface Surface {
  shape: MatrixShape
  area: LampArea
  count: number
  inverted: boolean
  mirrored: boolean
}

function surfaceOf(
  device: HardwareDeviceConfiguration,
  effect: LedEffect
): Surface | undefined {
  const area = areaOf(device, effect)
  if (!area) return undefined
  return {
    shape: deviceShape(device),
    area,
    count: area.width * area.height,
    inverted: effect.inverted ?? false,
    mirrored: effect.mirrored ?? false
  }
}

function surfaceSize(surface: Surface): number {
  return surface.mirrored ? Math.ceil(surface.count / 2) : surface.count
}

function place(frame: RgbColor[], surface: Surface, index: number, color: RgbColor): void {
  const put = (offset: number): void => {
    const column = offset % surface.area.width
    const row = Math.floor(offset / surface.area.width)
    if (!maskHolds(surface.area, column, row)) return
    const lamp = matrixLamp(surface.shape, surface.area.x + column, surface.area.y + row)
    if (lamp >= 0 && lamp < frame.length) frame[lamp] = color
  }
  if (!surface.mirrored) {
    put(surface.inverted ? surface.count - 1 - index : index)
    return
  }
  const half = surfaceSize(surface)
  put(surface.inverted ? index : half - 1 - index)
  put(surface.inverted ? surface.count - 1 - index : surface.count - half + index)
}

function shade(color: RgbColor, map: (channel: number) => number): RgbColor {
  const hex = (value: number): string =>
    Math.max(0, Math.min(255, value)).toString(16).padStart(2, '0')
  const channel = (at: number): number => Number.parseInt(color.slice(at, at + 2), 16)
  return `#${hex(map(channel(1)))}${hex(map(channel(3)))}${hex(map(channel(5)))}` as RgbColor
}

function dim(color: RgbColor, level: number): RgbColor {
  const clamped = Math.max(0, Math.min(1, level))
  return shade(color, (channel) => Math.trunc(channel * clamped))
}

function scaled(color: RgbColor, brightness: number): RgbColor {
  return shade(color, (channel) => Math.trunc((channel * brightness + 127) / 255))
}

function wheel(turn: number): RgbColor {
  const hue = (turn - Math.floor(turn)) * 6
  const sector = Math.floor(hue)
  const rise = Math.trunc((hue - sector) * 255)
  const fall = 255 - rise
  const hex = (r: number, g: number, b: number): RgbColor =>
    `#${[r, g, b].map((v) => v.toString(16).padStart(2, '0')).join('')}` as RgbColor
  switch (sector) {
    case 0: return hex(255, rise, 0)
    case 1: return hex(fall, 255, 0)
    case 2: return hex(0, 255, rise)
    case 3: return hex(0, fall, 255)
    case 4: return hex(rise, 0, 255)
    default: return hex(255, 0, fall)
  }
}

function paintEffect(
  frame: RgbColor[],
  device: HardwareDeviceConfiguration,
  effect: LedEffect,
  input: PaintInput,
  colors: LayerColors,
  value: number | undefined
): void {
  const surface = surfaceOf(device, effect)
  if (!surface) return
  const lamps = surfaceSize(surface)
  const color = colors.ink
  const fraction = value === undefined ? 0 : rangeFraction(value, effect.minimum, effect.maximum)
  const period = (effect.speed_ms ?? 1000) || 1000
  const phase = (input.elapsedMs % period) / period

  switch (effect.type ?? 'solid') {
    case 'solid':
      for (let index = 0; index < lamps; ++index) place(frame, surface, index, color)
      break
    case 'gradient':
      for (let index = 0; index < lamps; ++index) {
        const position = lamps <= 1 ? 0 : index / (lamps - 1)
        place(frame, surface, index, rampColor(effect.stops, position) ?? color)
      }
      break
    case 'steps': {
      const steps = effect.steps ?? []
      if (steps.length === 0) break
      for (let index = 0; index < lamps; ++index) {
        const step = steps[Math.min(Math.floor((index * steps.length) / lamps), steps.length - 1)]
        if (!step || fraction + 1e-6 < (step.threshold ?? 0)) continue
        place(frame, surface, index, step.color ?? color)
      }
      break
    }
    case 'gauge': {
      const filled = rampColor(effect.stops, value ?? 0) ?? color
      const lit = Math.round(fraction * lamps)
      for (let index = 0; index < lit && index < lamps; ++index) {
        place(frame, surface, index, filled)
      }
      break
    }
    case 'animation': {
      const head = phase * lamps
      const sweep = 1 - Math.abs(2 * phase - 1)
      const breath = 0.5 - 0.5 * Math.cos(phase * 2 * Math.PI)
      for (let index = 0; index < lamps; ++index) {
        switch (effect.animation ?? 'rainbow') {
          case 'rainbow':
            place(frame, surface, index, wheel(phase + index / lamps))
            break
          case 'pulse':
            place(frame, surface, index, dim(color, breath))
            break
          case 'wipe':
            if (index < head) place(frame, surface, index, color)
            break
          case 'chase':
            if (Math.floor(head) % lamps === index) place(frame, surface, index, color)
            break
          case 'scan':
            place(frame, surface, index, dim(color, 1 - Math.abs(index - sweep * (lamps - 1))))
            break
        }
      }
      break
    }
    default:
      break
  }
}

function paintMatrix(
  frame: RgbColor[],
  device: HardwareDeviceConfiguration,
  effect: LedEffect,
  input: PaintInput,
  colors: LayerColors,
  valueText: string | undefined,
  value: number | undefined
): void {
  if (!isMatrix(device)) return
  const panel = panelOf(device, frame, effect)
  if (!panel) return
  if ((effect.type ?? 'solid') === 'sprite') {
    paintSprite(panel, device, effect, value, colors.tint, input.elapsedMs)
    return
  }
  const shown = valueText ?? (value === undefined ? '' : wholeText(value))
  paintText(panel, effect, (effect.text ?? '') + shown, colors.ink, input.elapsedMs)
}

function lit(frame: RgbColor[], brightness: number): RgbColor[] {
  if (brightness >= FULL_BRIGHTNESS) return frame
  return frame.map((color) => (color === OFF ? color : scaled(color, brightness)))
}

export function paintOutput(
  device: HardwareDeviceConfiguration,
  input: PaintInput
): RgbColor[] {
  const lamps = lampsOf(device)
  const frame: RgbColor[] = new Array(lamps).fill(OFF)
  for (const [index, effect] of (device.effects ?? []).entries()) {
    if (input.gates[index] === false) continue
    const state = input.states?.[index] ?? newRuleState()
    const colors = colorsOf(effect, input.watched?.[index], state, input.elapsedMs)
    const blinkMs = colors.blinkMs ?? effect.blink_ms
    if (blinkMs) {
      const since = colors.blinkMs ? (colors.sinceMs ?? 0) : 0
      const half = blinkMs / 2
      if (Math.floor((input.elapsedMs - since) / half) % 2 === 1) continue
    }
    const area = colors.background ? areaOf(device, effect) : undefined
    if (area && colors.background) fillArea(frame, device, area, colors.background)
    const value = valueOf(effect, input.values?.[index], input.value)
    const type = effect.type ?? 'solid'
    if (type === 'sprite' || type === 'text') {
      paintMatrix(frame, device, effect, input, colors, input.valueText?.[index], value)
      continue
    }
    paintEffect(frame, device, effect, input, colors, value)
  }
  return lit(frame, device.brightness ?? DEFAULT_BRIGHTNESS)
}

File: firmware/modules/rgb_leds/src/led_paint.cpp
#include "led_paint.hpp"

#include <algorithm>
#include <cmath>
#include <numbers>

#include "value_conditions.hpp"

namespace pitrig::rgb_leds {
namespace {

using configuration::LedEffect;

[[nodiscard]] led::Color ramp_at(const LedEffect& effect, const float position,
                                 const led::Color fallback) {
  const std::optional<std::uint32_t> color = conditions::ramp_color(
      {effect.stops.data(), effect.stop_count}, static_cast<double>(position));
  return color.has_value() ? led::Color::from_rgb(*color) : fallback;
}

void paint_solid(const Surface& surface, const led::Color color) {
  for (std::size_t index = 0; index < surface.size(); ++index) {
    surface.set(index, color);
  }
}

void paint_gradient(const Surface& surface, const LedEffect& effect,
                    const led::Color fallback) {
  const std::size_t lamps = surface.size();
  for (std::size_t index = 0; index < lamps; ++index) {
    const float position =
        lamps <= 1 ? 0.0F : static_cast<float>(index) / (lamps - 1);
    surface.set(index, ramp_at(effect, position, fallback));
  }
}

void paint_steps(const Surface& surface, const LedEffect& effect,
                 const float fraction) {
  const std::size_t lamps = surface.size();
  if (effect.step_count == 0 || lamps == 0) {
    return;
  }
  for (std::size_t index = 0; index < lamps; ++index) {
    const std::size_t step = std::min<std::size_t>(
        index * effect.step_count / lamps, effect.step_count - 1U);
    if (fraction + 1e-6F < effect.steps[step].threshold) {
      continue;
    }
    surface.set(index, led::Color::from_rgb(effect.steps[step].color));
  }
}

void paint_gauge(const Surface& surface, const LedEffect& effect,
                 const led::Color fallback, const float fraction,
                 const float value) {
  const led::Color color =
      effect.stop_count >= 2 ? ramp_at(effect, value, fallback) : fallback;
  const std::size_t lamps = surface.size();
  const auto lit =
      static_cast<std::size_t>(std::lround(fraction * lamps));
  for (std::size_t index = 0; index < lit && index < lamps; ++index) {
    surface.set(index, color);
  }
}

[[nodiscard]] led::Color wheel(const float turn) {
  const float hue = (turn - std::floor(turn)) * 6.0F;
  const auto sector = static_cast<int>(hue);
  const auto rise = static_cast<std::uint8_t>((hue - sector) * 255.0F);
  const auto fall = static_cast<std::uint8_t>(255 - rise);
  switch (sector) {
    case 0:
      return {255, rise, 0};
    case 1:
      return {fall, 255, 0};
    case 2:
      return {0, 255, rise};
    case 3:
      return {0, fall, 255};
    case 4:
      return {rise, 0, 255};
    default:
      return {255, 0, fall};
  }
}

void paint_animation(const Surface& surface, const LedEffect& effect,
                     const led::Color color, const float phase) {
  const std::size_t lamps = surface.size();
  if (lamps == 0) {
    return;
  }
  const float head = phase * lamps;
  const float sweep = 1.0F - std::abs(2.0F * phase - 1.0F);
  const float breath =
      0.5F - 0.5F * std::cos(phase * 2.0F * std::numbers::pi_v<float>);
  for (std::size_t index = 0; index < lamps; ++index) {
    const float position = static_cast<float>(index) / lamps;
    switch (effect.animation) {
      case configuration::LedAnimationKind::rainbow:
        surface.set(index, wheel(phase + position));
        break;
      case configuration::LedAnimationKind::pulse:
        surface.set(index, color.faded(breath));
        break;
      case configuration::LedAnimationKind::wipe:
        if (static_cast<float>(index) < head) {
          surface.set(index, color);
        }
        break;
      case configuration::LedAnimationKind::chase:
        if (static_cast<std::size_t>(head) % lamps == index) {
          surface.set(index, color);
        }
        break;
      case configuration::LedAnimationKind::scan: {
        const float distance =
            std::abs(static_cast<float>(index) - sweep * (lamps - 1));
        surface.set(index, color.faded(1.0F - distance));
        break;
      }
    }
  }
}

[[nodiscard]] int matching_rule(const LedEffect& effect,
                               const std::optional<double> watched) {
  if (!watched.has_value()) {
    return -1;
  }
  for (std::uint8_t index = 0; index < effect.color_rule_count; ++index) {
    const configuration::LedColorRule& rule = effect.color_rules[index];
    if (conditions::condition_holds(rule.op, *watched, rule.value)) {
      return index;
    }
  }
  return -1;
}

[[nodiscard]] int applied_rule(const LedEffect& effect,
                               const std::optional<double> watched,
                               ColorRuleState& state,
                               const std::uint64_t now_us) {
  int applied = matching_rule(effect, watched);
  if (applied >= 0) {
    state.hold_until_us =
        now_us +
        static_cast<std::uint64_t>(effect.color_rules[applied].hold_ms) * 1000;
  } else if (state.applied >= 0 && state.applied < effect.color_rule_count &&
             now_us < state.hold_until_us) {
    applied = state.applied;
  }
  if (applied != state.applied) {
    state.applied = applied;
    state.started_us = now_us;
  }
  return applied;
}

}

LayerColors colors_of(const configuration::LedEffect& effect,
                      const std::optional<double> watched,
                      ColorRuleState& state, const std::uint64_t now_us) {
  LayerColors colors{.ink = led::Color::from_rgb(effect.color)};
  if (effect.background_color != configuration::kTransparentColor) {
    colors.background = led::Color::from_rgb(effect.background_color);
  }
  const int applied = applied_rule(effect, watched, state, now_us);
  if (applied < 0) {
    return colors;
  }
  const configuration::LedColorRule& rule = effect.color_rules[applied];
  if (rule.color != configuration::kTransparentColor) {
    colors.tint = led::Color::from_rgb(rule.color);
    colors.ink = *colors.tint;
  }
  if (rule.background_color != configuration::kTransparentColor) {
    colors.background = led::Color::from_rgb(rule.background_color);
  }
  colors.blink_ms = rule.blink_ms;
  colors.since_us = state.started_us;
  return colors;
}

void paint(const Surface& surface, const configuration::LedEffect& effect,
           const led::Color color, const std::optional<double> value,
           const std::uint64_t elapsed_us) {
  const float fraction =
      value.has_value() ? conditions::range_fraction(*value, effect.range)
                        : 0.0F;
  const std::uint64_t period_us =
      static_cast<std::uint64_t>(effect.speed_ms == 0 ? 1000
                                                      : effect.speed_ms) *
      1000;
  const float phase = static_cast<float>(elapsed_us % period_us) /
                      static_cast<float>(period_us);

  switch (effect.type) {
    case configuration::LedEffectType::solid:
      paint_solid(surface, color);
      break;
    case configuration::LedEffectType::gradient:
      paint_gradient(surface, effect, color);
      break;
    case configuration::LedEffectType::steps:
      paint_steps(surface, effect, fraction);
      break;
    case configuration::LedEffectType::gauge:
      paint_gauge(surface, effect, color, fraction,
                  static_cast<float>(value.value_or(0.0)));
      break;
    case configuration::LedEffectType::animation:
      paint_animation(surface, effect, color, phase);
      break;
    case configuration::LedEffectType::sprite:
    case configuration::LedEffectType::text:
      break;
  }
}

}
